libs/types-common/src/Data/Qualified.hs

```diff
@@ -4,18 +4,54 @@ module Data.Qualified where
 
 import Data.Aeson (FromJSON, ToJSON, withText)
 import qualified Data.Aeson as Aeson
+import Data.Attoparsec.ByteString (takeByteString)
 import Data.Bifunctor (first)
 import qualified Data.ByteString.Conversion as BS.C
+import Data.ByteString.Conversion (FromByteString (parser))
 import Data.Domain (Domain, domainText, mkDomain)
 import Data.Handle (Handle (..))
 import Data.Id (Id (toUUID))
 import Data.String.Conversions (cs)
 import qualified Data.Text as Text
+import qualified Data.Text.Encoding as Text.E
 import qualified Data.UUID as UUID
 import Imports hiding (local)
 import Servant.API (FromHttpApiData (parseUrlPiece))
 import Test.QuickCheck (Arbitrary (arbitrary))
 
+----------------------------------------------------------------------
+-- OPTIONALLY QUALIFIED
+
+type OptionallyQualified a = Either a (Qualified a)
+
+unqualified :: a -> OptionallyQualified a
+unqualified = Left
+
+qualified :: Qualified a -> OptionallyQualified a
+qualified = Right
+
+{-
+TODO: do it properly
+Either is a hack, conceptually we want this:
+
+data OptionallyQualified a
+  = OptionallyQualified
```

Comment:
Do you still want to resolve htis before merging?

Reply:
Yes, that's the main things that's missing.

Reply:
Implemented now.